src/main/java/co/cask/gcp/gcs/StorageClient.java

```diff
@@ -0,0 +1,257 @@
+  private final Storage storage;
+
+  private StorageClient(Storage storage) {
+    this.storage = storage;
+  }
+
+  /**
+   * Copy objects from the source path to the destination path. If the source path is a single object, that object
+   * will be copied to the destination. If the source path represents a directory, objects within the directory
+   * will be copied to the destination directory.
+   *
+   * @param sourcePath the path to copy objects from
+   * @param destPath the path to copy objects to
+   * @param recursive whether to copy objects in all subdirectories
+   * @param overwrite whether to overwrite existing objects
+   * @throws IllegalArgumentException if overwrite is false and copying would overwrite an existing object
+   */
+  public void copy(String sourcePath, String destPath, boolean recursive, boolean overwrite) {
+    pairTraverse(sourcePath, destPath, recursive, overwrite, BlobPair::copy);
+  }
+
+  /**
+   * Move objects from the source path to the destination path. If the source path is a single object, that object
+   * will be moved to the destination. If the source path represents a directory, objects within the directory
+   * will be moved to the destination directory.
+   *
+   * @param sourcePath the path to move objects from
+   * @param destPath the path to move objects to
+   * @param recursive whether to move objects in all subdirectories
+   * @param overwrite whether to overwrite existing objects
+   * @throws IllegalArgumentException if overwrite is false and moving would overwrite an existing object
+   */
+  public void move(String sourcePath, String destPath, boolean recursive, boolean overwrite) {
+    pairTraverse(sourcePath, destPath, recursive, overwrite, BlobPair::move);
+  }
+
+  /**
+   * Gets source and destination pairs by traversing the source path. Consumes each pair after the directory structure
+   * is completely traversed.
+   */
+  private void pairTraverse(String sourcePath, String destPath, boolean recursive, boolean overwrite,
+                            Consumer<BlobPair> consumer) {
+    BlobId baseSourceBlobID = GCSConfigHelper.getBlobId(sourcePath);
+    BlobId baseDestBlobID = GCSConfigHelper.getBlobId(destPath);
+
+    Bucket sourceBucket = storage.get(baseSourceBlobID.getBucket());
+    if (sourceBucket == null) {
+      throw new IllegalArgumentException(
+        String.format("Source bucket '%s' does not exist.", baseSourceBlobID.getBucket()));
+    }
+    Bucket destBucket = storage.get(baseDestBlobID.getBucket());
+    if (destBucket == null) {
+      throw new IllegalArgumentException(
+        String.format("Destination bucket '%s' does not exist. Please create it first.", baseSourceBlobID.getBucket()));
+    }
+
+    boolean destinationBaseExists;
+    String baseDestName = baseDestBlobID.getName();
+    if (baseDestBlobID.getName().isEmpty()) {
+      // if the name is empty, it represents a bucket and not a blob.
```

Comment:
do you think it's a good idea to encapsulate this logic somehow instead of relying on how getBlobId works?

Reply:
I agree it would be cleaner, the current way kind of abuses the BlobId class.

Reply:
moved this to a GCSPath class, which replaces the GCSConfigHelper class.